main/src/ca/uwaterloo/flix/api/Flix.scala

```diff
@@ -509,31 +509,52 @@ class Flix {
       implicit def map[C](f: A => C): Validation[C, B] = Validation.mapN(v)(f)
     }
 
+    val errors = mutable.ListBuffer.empty[CompilationMessage]
+
     val (afterReader, readerErrors) = Reader.run(getInputs, knownClassesAndInterfaces)
+    errors ++= readerErrors
+
     val (afterLexer, lexerErrors) = Lexer.run(afterReader, cachedLexerTokens, changeSet)
+    errors ++= lexerErrors
+
     val (afterParser, parserErrors) = Parser2.run(afterLexer, cachedParserCst, changeSet)
+    errors ++= parserErrors
+
+    val (weederValidation, weederErrors) = Weeder2.run(afterReader, entryPoint, afterParser, cachedWeederAst, changeSet)
+    errors ++= weederErrors
 
     /** Remember to update [[AstPrinter]] about the list of phases. */
     val result = for {
-      afterWeeder <- Weeder2.run(afterReader, entryPoint, afterParser, cachedWeederAst, changeSet).withSoftFailures(readerErrors).withSoftFailures(lexerErrors).withSoftFailures(parserErrors)
+      afterWeeder <- weederValidation
       afterDesugar = Desugar.run(afterWeeder, cachedDesugarAst, changeSet)
-      (afterNamer, namerErrors) = Namer.run(afterDesugar)
+      (afterNamer, nameErrors) = Namer.run(afterDesugar)
+      _ = errors ++= nameErrors
```

Comment:
What is this magic?... Ah its because its in a for-comprehension. But perhaps we should just call weeder and pattern match? And then again for Resolver. Yes its two layers of indentation, but it feels more clear to me, and then we can also drop the monadic stuff from Validation (AsMonad)

Reply:
Ah- lets do it in the next then 👍